src/backend/distributed/utils/background_jobs.c

```diff
@@ -385,292 +648,339 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
+
+
+						/*
+						 * reload task while holding a new ExclusiveLock on the table. A separate process
+						 * could have cancelled or removed the task by now, they would not see the pid and
+						 * status update, so it is our responsibility to stop the backend and _not_ write
+						 * the pid and running status.
+						 *
+						 * The lock will release on transaction commit.
+						 */
+						LockRelationOid(DistBackgroundTaskRelationId(), ExclusiveLock);
+
+						task = GetBackgroundTaskByTaskId(handleEntry->taskid);
+
+						ereport(LOG, (errmsg("found task with jobid/taskid: %ld/%ld",
+											 task->jobid, task->taskid)));
+
+						if (!task || task->status == BACKGROUND_TASK_STATUS_CANCELLING)
+						{
+							/*
+							 * being in that step means that a concurrent cancel happened. we should make task status
+							 * as cancelled. We also want to reflect cancel message by consuming task executor
+							 * queue.
+							 */
+							taskCancelledConcurrently = true;
+						}
+						else
+						{
+							/*
+							 * now that we have verified the task has not been cancelled and still exist we
+							 * update it to reflect the new state. If task is already in running status,
+							 * the operation is idempotent. But for runnable tasks, we make their status
+							 * as running.
+							 */
+
+							pid_t pid = 0;
+							GetBackgroundWorkerPid(handleEntry->handle, &pid);
+							task->status = BACKGROUND_TASK_STATUS_RUNNING;
+							SET_NULLABLE_FIELD(task, pid, pid);
+
+							/* Update task status to indicate it is running */
+							UpdateBackgroundTask(task);
+							UpdateBackgroundJob(task->jobid);
+						}
+
+						monitorExecutionState = TryConsumeTaskWorker;
+						break;
+					}
+
+					case TryConsumeTaskWorker:
+					{
+						/*
+						 * we consume task executor response queue.
+						 * possible response codes can lead us different steps below.
+						 */
+						bool hadError = false;
+						shm_mq_result mq_res = ReadFromExecutorQueue(handleEntry->seg,
+																	 handleEntry->message,
+																	 &hadError);
+
+						if (taskCancelledConcurrently)
+						{
+							/* we update task status as cancelled */
+							ereport(LOG, (errmsg(
+											  "task jobid/taskid is cancelled: %ld/%ld",
+											  task->jobid, task->taskid)));
+
+							task->status = BACKGROUND_TASK_STATUS_CANCELLED;
+							monitorExecutionState = TaskEnded;
+						}
+						else if (hadError)
+						{
+							ereport(LOG, (errmsg("task jobid/taskid failed: %ld/%ld",
+												 task->jobid, task->taskid)));
+
+							monitorExecutionState = TaskHadError;
+						}
+						else if (mq_res == SHM_MQ_WOULD_BLOCK)
+						{
+							/*
+							 * still running the task but had no output for current iteration.
+							 * commit current transaction and jump onto next task.
+							 */
+							PopActiveSnapshot();
+							CommitTransactionCommand();
+							monitorExecutionState = ExecutionEnded;
```

Comment:
the ExecutionEnded is confusing to me. This seems more like a TaskStillRunning state?

Reply:
I removed ExecutionEnded state, instead used readable flags representing loop breaking conditions.